rust/capture/src/passes/invalid_surrogates.rs

```diff
@@ -0,0 +1,353 @@
+        assert_eq!(data, r#"{"event":"\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_unpaired_low_surrogate() {
+        let raw_data = r#"{"event":"\uDC00"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_wrong_order_surrogate_pair() {
+        let raw_data = r#"{"event":"\uDC00\uD800"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_trailing_escape() {
+        let raw_data = r#"{"event":"\u"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_trailing_escape_pair() {
+        let raw_data = r#"{"event":"\u\u"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_trailing_escape_pair_high_surrogate() {
+        let raw_data = r#"{"event":"\uD800\u"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_trailing_escape_pair_low_surrogate() {
+        let raw_data = r#"{"event":"\uDC00\u"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD\uFFFD"}"#);
+    }
+
+    #[test]
+    fn test_trailing_escape_char() {
+        let raw_data = r#"{"event":"\uD800\"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uFFFD\"#);
+    }
+
+    #[test]
+    fn test_valid_pair_trailing_slash() {
+        let raw_data = r#"{"event":"\uD800\uDC00\"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uD800\uDC00\"#);
+    }
+
+    #[test]
+    fn it_handles_normal_data() {
+        let raw_data = r#"{"event":"\uD800\uDC00"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\uD800\uDC00"}"#);
+    }
+
+    #[test]
+    fn it_handles_data_with_no_surrogates() {
+        let raw_data = r#"{"event":"\n"}"#;
+        let pass = super::InvalidSurrogatesPass::new(raw_data.chars());
+        let data = pass.collect::<String>();
+        assert_eq!(data, r#"{"event":"\n"}"#);
+    }
+
+    #[test]
+    fn it_handles_actual_session_data() {
```

Comment:
And then this test is based on my own session replay data that I intercepted while browsing our site - it would have triggered the bug that caused the outage last time, and I've confirmed the asserts below fail without the fix above, and pass with it